xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/doc/XWikiDocument.java

```diff
@@ -492,9 +493,7 @@ public String toString()
      */
     private Map<DocumentReference, List<BaseObject>> xObjects = new TreeMap<DocumentReference, List<BaseObject>>();
 
-    // TODO: use a Map instead of a List to store attachment in XWikiDocument
-    private final List<XWikiAttachment> attachmentList =
-        new AbstractNotifyOnUpdateList<XWikiAttachment>(new ArrayList<XWikiAttachment>())
+    private final List<XWikiAttachment> attachmentList = new XWikiAttachmentList()
```

Comment:
You should make attachmentList be a XWikiAttachmentList, that would remove all the casts you are currently doing.